[wasm] Add support for symbolicating native traces from JS, using a symbols file

radical opened the pull request:
- adds support for symbolicating error, and warning messages
- `debug.js.symbols` is loaded from the vfs
- This is enabled by default, only for non-aot tests (since we don't strip aot tests)
- Works for debugger tests too

radical committed:
[wasm] Fix run command lines for samples

radical committed:
[wasm] Export FS_readFile

radical committed:
[wasm] Add support for symbolicating traces in console errors, or

.. warnings. The regex patterns are hardcoded in `debug.ts`. And the
symbols are loaded from `dotnet.js.symbols`.

We try to symbolicate wherever possible.

radical committed:
[wasm] Enable symbolicating for non-aot test builds by default

radical commented:
/azp run runtime-wasm

radical committed:
[wasm] Enable symbol map for debugger tests

pavelsavara commented on src/mono/wasm/runtime/debug.ts:
```diff
@@ -314,25 +334,96 @@ export function mono_wasm_debugger_log(level: number, message_ptr: CharPtr): voi
 }
 
+function _readSymbolMapFile(filename: string): void {
+    if (Module.FS_readFile === undefined)
```
That should not really happen, I think.

pavelsavara commented:
In the future we could also add symbols to stack traces which are marshaled as C# exception.
See https://github.com/dotnet/runtime/blob/842f4d76a8d9344fa6d2e7dbd1534b7b7a6fec57/src/mono/wasm/runtime/method-calls.ts#L459

pavelsavara commented on src/mono/wasm/runtime/types/emscripten.ts:
```diff
@@ -49,6 +49,7 @@ export declare interface EmscriptenModule {
     UTF8ArrayToString(u8Array: Uint8Array, idx?: number, maxBytesToRead?: number): string;
     FS_createPath(parent: string, path: string, canRead?: boolean, canWrite?: boolean): string;
     FS_createDataFile(parent: string, name: string, data: TypedArray, canRead: boolean, canWrite: boolean, canOwn?: boolean): string;
+    FS_readFile(filename: string, opts: any): any;
```
this should force change in `dotnet.d.ts`

radical committed:
Address feedback from @pavelsavara

radical committed:
[wasm] Add a `mono_wasm_symbolicate_string` function

.. and use that to symbolicate traces marshaled as C# exceptions (thanks
to @pavelsavara for the suggestion).

radical committed:
Merge remote-tracking branch 'origin/main' into symbol-map-2

radical commented:
/azp run runtime-wasm

radical committed:
Merge branch 'dotnet:main' into symbolicate-from-js

radical commented:
Test failures are unrelated.

radical committed:
Merge branch 'dotnet:main' into symbolicate-from-js

radical commented:
Failures are https://github.com/dotnet/runtime/issues/66231, and https://github.com/dotnet/runtime/issues/65817 .

radekdoulik commented on src/mono/sample/wasm/wasm.mk:
```diff
@@ -39,7 +39,7 @@ run-browser:
 	fi
 
 run-console:
-	cd bin/$(CONFIG)/AppBundle && $(V8_PATH) --stack-trace-limit=1000 --single-threaded --expose_wasm $(MAIN_JS) -- $(DOTNET_MONO_LOG_LEVEL) --run $(CONSOLE_DLL) $(ARGS)
```
This broke browser-bench's `make run` for console version. Why was `--run $(CONSOLE_DLL)` removed? Do I need to update browser-bench's Makefile somehow?
